### DISCUSSION TITLE
API Client?

### DISCUSSION THREAD

**Discussion by mastercoms** [piccolo-orm/piccolo#488] [General]
Does an API web client exist for building REST queries for PiccoloCRUD? I find the syntax a bit unwieldy for production, so we'd make a query builder for this, if one does not already exist.

**Comment by dantownsend** [comment]
@mastercoms There isn't one yet, but it's a nice idea.

If you're interested in open sourcing it at some point, let me know if I can help at all.

**Comment by mastercoms** [comment]
https://github.com/OpenAPITools/openapi-generator/

Seems like this can be used

**Reply by dantownsend** [reply]
Have you seen this one as well?

https://github.com/swagger-api/swagger-codegen

It's pretty similar - I'm using it at the moment to generate JS client libraries, and does a decent job:

**Reply by mastercoms** [reply]
Haven't, will check it out thanks!

**Comment by mastercoms** [comment]
It works decently but have a problem with large tables where the API calls have a lot of parameters in TypeScript. Do you know of a way to have a params object instead, I'm not sure if it's best done in adjusting the REST call to use JSON body instead of query params, or to do it in the client generator.

**Reply by dantownsend** [reply]
It just works for GET params at the moment. Is the problem with TypeScript that it won't allow a URL above a certain length?

**Reply by mastercoms** [reply]
It was just the query params giving me trouble on swagger codegen, openapi generator seems to have an option to make it simpler

https://github.com/codeserk/openapi-generator/blob/master/docs/generators/typescript-axios.md

useSingleRequestParameter

**Comment by mastercoms** [comment]
Is there a way to automatically join tables together instead of querying the schema and then looping through all possible visible fields?

**Reply by dantownsend** [reply]
I don't think so at the moment - what API did you have in mind? Some kind of GET param?

**Reply by mastercoms** [reply]
I think there's a few options or things that could be done:

* Specify visible tables (including self) so that visible fields can act as a purely additive selection for joins
* Specify field _exclusions_ rather than just the inclusive option we have now
* Specify default visible fields / default excluded fields on the API side
* Specify a join depth for automatically including fields

And as another note, I think the current query parameter for visible fields (comma separated string) has issues with validation. If it would be possible to make it a Swagger array parameter instead, maybe we could get more information through about the available options that can be passed in the array.

**Reply by dantownsend** [reply]
Lots of interesting ideas there.

It seems like the crux of the problem is if you have a large table with lots of foreign keys, then using `visible_fields` becomes a bit tedious, and the URL gets really long.

We get around this in Piccolo usually with `all_columns`:

```python
await Band.select(Band.all_columns(), Band.manager.all_columns())
```

Maybe we need a similar concert for `visible_fields`. Something like this?

```
/band/?__visible_fields=_all,manager._all
```

I can see the value in having `excluded_fields` too.

> And as another note, I think the current query parameter for visible fields (comma separated string) has issues with validation. If it would be possible to make it a Swagger array parameter instead, maybe we could get more information through about the available options that can be passed in the array.

For `visible_columns` we're just doing this at the moment to validate it's a correct column name:

https://github.com/piccolo-orm/piccolo_api/blob/b0c6c86e074044c33b648a25114b053e0016fcdd/piccolo_api/crud/endpoints.py#L1023

There might be some way to break it - I'm not sure.

**Reply by mastercoms** [reply]
> It seems like the crux of the problem is if you have a large table with lots of foreign keys, then using visible_fields becomes a bit tedious, and the URL gets really long.

Yep, that's the problem.

> Maybe we need a similar concert for visible_fields. Something like this?

That would be a great solution!
